[Store] Add Redis

lyrixx opened the pull request:
| Q             | A
| ------------- | ---
| Bug fix?      | no
| New feature?  | yes
| Docs?         | no
| Issues        |
| License       | MIT

I continue to bench different stores, and the next one is Redis

## Demo

```console
$ php rag/redis.php                                                                                                                   
Here are some movies that fit the theme of technology:                                                                                                                                      
                                                                                                                                                                                            
1. **The Matrix**                                                                                                                                                                           
   - **Description**: A hacker discovers the world he lives in is a simulated reality and joins a rebellion to overthrow its controllers.                                                   
   - **Director**: The Wachowskis                                                                                                                                                           
                                                                                                                                                                                            
2. **Interstellar**                                                                                                                                                                         
   - **Description**: A team of explorers travel through a wormhole in space in an attempt to ensure humanity's survival.                                                                   
   - **Director**: Christopher Nolan                                                                                                                                                        
                                                                                                                                                                                            
3. **Jurassic Park**                                                                                                                                                                        
   - **Description**: During a preview tour, a theme park suffers a major power breakdown that allows its cloned dinosaur exhibits to run amok.                                             
   - **Director**: Steven Spielberg
```


---

## Symfony configuration

### With a custom redis connection


```yaml
ai:
    store:
        redis:
            name:
                # a service id of a Redis client
                client: my_redis
                index_name: "my_index" # required
                key_prefix: 'vector:'

                # Distance metric to use for vector similarity search
                distance: !php/enum Symfony\AI\Store\Bridge\Redis\Distance::Cosine

services:
    my_redis:
        class: Redis
        arguments:
            $options:
                host: '127.0.0.1'
```

### Without a custom redis connection


```yaml
ai:
    store:
        redis:
            name:
                # see https://github.com/phpredis/phpredis?tab=readme-ov-file#example-1
                connection_parameters:
                    host: 127.0.0.1
                    port: 6379

                index_name: "my_index"
```

GromNaN commented on src/ai-bundle/config/options.php:
```diff
@@ -252,10 +253,30 @@
+                        ->useAttributeAsKey('name')
+                        ->arrayPrototype()
+                            ->children()
+                                ->variableNode('connection_parameters')
```
It's good that the `RedisStore` only takes an instance of `Redis`. It is not its responsibility to instantiate the Redis client.
The bundle configuration should accept a redis service name, which could be created directly in the application or with SncRedisBundle (which [accepts dsn in config](https://github.com/snc/SncRedisBundle/blob/master/docs/README.md#usage)).

lyrixx replied:
I update the config to support either 
* An array of parameter that is passed to the Redis constructor
* A service id that points to an instance of Redis

chr-hertel commented:
Could you bring in a rag example as well, please? redis cloud or docker compose maybe

OskarStark commented:
docker compose if possible please 🙏

lyrixx commented:
I added a rag example, with docker compose. I also updated the PR description

chr-hertel commented:
Needs a rebase after #317

OskarStark commented:
Open to finish this PR?

lyrixx commented:
Oupsy, I forgot this one. I'll finish it ASAP

lyrixx commented on examples/rag/redis.php:
```diff
@@ -0,0 +1,76 @@
+$indexer = new Indexer(new InMemoryLoader($documents), $vectorizer, $store, logger: logger());
+$indexer->index($documents);
+
+$model = new Gpt(Gpt::GPT_4O_MINI);
```
For the record, I copied the content from `examples/rag/qdrant.php`

OskarStark commented on src/store/src/Bridge/Redis/Store.php:
```diff
@@ -0,0 +1,184 @@
+    /**
+     * @param array{vector_size?: positive-int, index_method?: string, extra_schema?: list<string>} $options
+     *
+     * - For Mistral: ['vector_size' => 1024]
+     * - For OpenAI: ['vector_size' => 1536]
+     * - For Gemini: ['vector_size' => 3072] (default)
+     * - For other models: adjust vector_size accordingly
```
lets remove this

OskarStark commented on src/ai-bundle/src/AiBundle.php:
```diff
@@ -750,8 +750,8 @@ private function processStoreConfig(string $type, array $stores, ContainerBuilde
             foreach ($stores as $name => $store) {
-                if (isset($store['http_client'])) {
-                    $httpClient = new Reference($store['http_client']);
+                if (isset($store['client'])) {
+                    $httpClient = new Reference($store['client']);
```
needed?

lyrixx replied:
rah, that's a typo. Sorry. (I understand now why I had to change twice a piece of code 🤦🏼, haha)

lyrixx commented on src/store/src/Bridge/Redis/Store.php:
```diff
@@ -0,0 +1,184 @@
+    public function __construct(
+        private readonly \Redis $redis,
+        private readonly string $indexName,
+        private readonly string $keyPrefix = 'vector:',
```
It's not needed, but it's a common practice to "namespace" redis keys

lyrixx commented on src/store/composer.json:
```diff
@@ -38,6 +38,7 @@
         "symfony/ai-platform": "@dev",
         "symfony/clock": "^7.3|^8.0",
         "symfony/http-client": "^7.3|^8.0",
+        "symfony/polyfill-php83": "^1.32",
```
For `json_validate()` https://www.php.net/json_validate

lyrixx reviewed (commented):
Thanks for the review

OskarStark commented on src/store/src/Bridge/Redis/Store.php:
```diff
@@ -0,0 +1,183 @@
+    public function setup(array $options = []): void
+    {
+        $vectorSize = $options['vector_size'] ?? 3072;
+        $indexMethod = $options['index_method'] ?? 'FLAT'; // Or 'HNSW' for approximate search
```
if only `FLAT` and `HNSW` is allowed, I would add that to the array shape explicitly

lyrixx replied:
Can you check if I did it as you expected?

OskarStark commented on src/store/src/Bridge/Redis/Store.php:
```diff
@@ -0,0 +1,183 @@
+    {
+        $vectorSize = $options['vector_size'] ?? 3072;
+        $indexMethod = $options['index_method'] ?? 'FLAT'; // Or 'HNSW' for approximate search
+        $distanceMetric = $this->distance->getRedisMetric();
+        $extraSchema = $options['extra_schema'] ?? [];
+
+        // Create the index with vector field for JSON documents
+        try {
+            $this->redis->rawCommand(
+                'FT.CREATE', $this->indexName, 'ON', 'JSON',
+                'PREFIX', '1', $this->keyPrefix,
+                'SCHEMA',
+                '$.id', 'AS', 'id', 'TEXT',
+                '$.embedding', 'AS', 'embedding', 'VECTOR', $indexMethod, '6', 'TYPE', 'FLOAT32', 'DIM', $vectorSize, 'DISTANCE_METRIC', $distanceMetric,
+                ...$extraSchema,
```
```suggestion
        $extraSchema = $options['extra_schema'] ?? [];

        // Create the index with vector field for JSON documents
        try {
            $this->redis->rawCommand(
                'FT.CREATE', $this->indexName, 'ON', 'JSON',
                'PREFIX', '1', $this->keyPrefix,
                'SCHEMA',
                '$.id', 'AS', 'id', 'TEXT',
                '$.embedding', 'AS', 'embedding', 'VECTOR', $indexMethod, '6', 'TYPE', 'FLOAT32', 'DIM', $vectorSize, 'DISTANCE_METRIC', $this->distance->getRedisMetric(),
                ...$extraSchema,
```

OskarStark commented on src/store/src/Bridge/Redis/Distance.php:
```diff
@@ -0,0 +1,31 @@
+    case L2 = 'L2';
+    case Ip = 'IP';
+
+    public function getRedisMetric(): string
```
```suggestion
    public function getMetric(): string
```

It is already in the Redis namespace

OskarStark replied:
or shall we use `->value` instead of a getter?

OskarStark commented on src/store/src/Bridge/Redis/Store.php:
```diff
@@ -0,0 +1,183 @@
+    }
+
+    /**
+     * @param array<string, mixed> $options
```
can we be more specific for the array shape?

OskarStark reviewed (approved):
Some minors, afterwards good to merge

OskarStark commented on src/store/src/Bridge/Redis/Store.php:
```diff
@@ -0,0 +1,183 @@
+
+    /**
+     * @param array{vector_size?: positive-int, index_method?: string, extra_schema?: list<string>} $options
+     *
+     * - For Mistral: ['vector_size' => 1024]
+     * - For OpenAI: ['vector_size' => 1536]
+     * - For Gemini: ['vector_size' => 3072] (default)
```
```suggestion
```

lyrixx replied:
Why? It's super useful

OskarStark replied:
I think it's not possible to maintain this in the long run, but fine to keep it for now

OskarStark commented on src/store/src/Bridge/Redis/Store.php:
```diff
@@ -0,0 +1,183 @@
+                ...$extraSchema,
+            );
+        } catch (\RedisException $e) {
+            // Index might already exist, check if it's a "already exists" error
```
```suggestion
```

OskarStark commented:
Lets finish this one @lyrixx 🚀 👏

lyrixx commented on src/ai-bundle/src/AiBundle.php:
```diff
@@ -47,7 +47,6 @@
 use Symfony\AI\Platform\Bridge\VertexAi\PlatformFactory as VertexAiPlatformFactory;
 use Symfony\AI\Platform\Bridge\Voyage\PlatformFactory as VoyagePlatformFactory;
 use Symfony\AI\Platform\Exception\RuntimeException;
-use Symfony\AI\Platform\Model;
```
It's not used, isn't?

lyrixx commented:
Here we go 🚀 !

I hope and think I addressed all comments.
Tests are all green 🎉 

Thanks for the review, and sorry for the slow tempo on this one!

OskarStark commented:
It took time, but here we go, this is in now. Thank you very much @lyrixx.
